Requestor: Paloma Gupta
Requestee: Ibrahim El-Amin
RequestOrReplied: Replied
Must-fix: None
Verified:
- revert->red: in `stage_install`'s `except` I replaced the `_rollback(target, orig, plan.branch)` call with `pass` (rollback neutralized). `framework/tests/test_install_branch.py::test_failed_install_rolls_back` went RED: `AssertionError: assert 'real-team/install' == 'main'` — with rollback gone the tree is stranded on the install branch with the half-run install's `.claude/partial` debris still present, exactly what the guard must prevent. Restored the call -> 1 passed; `git diff` on the file is clean.
- refuse-not-fallback + compose-with-manifest: confirmed both refusal paths emit an actionable message and create nothing. Live non-git run prints "not a git repository ... Use a plain install and `restore` to undo it" and exits 1; `test_dirty_tree_refuses`/`test_non_git_repo_refuses`/`test_branch_collision_refuses` assert `staged is False` + no branch created. Composition verified by `test_composes_with_manifest_restore`: the installer's archive manifest + `CLAUDE.md.bak` are captured in the branch commit AND `restore.plan_restore` on the staged branch still reports `archived == ['.claude']` and reverts `['CLAUDE.md']` — the two undo paths coexist. `restore.py` and `node/` are untouched (file-disjoint with S2 confirmed via `git diff --name-only`).
- rollback / collision / consent-safety: `_rollback` does `checkout --force` + `clean -fd` + `branch -D`; the `clean -fd` is safe because a dirty tree is refused upfront, so untracked-at-rollback is only install debris (verified: post-fail tree has no `.claude`, not dirty, back on orig). Empty-install path also rolls back to avoid a dangling branch. Collision refused (`test_branch_collision_refuses`). Non-TTY without `--non-interactive` refuses at the consent gate, exit 1, nothing created (`test_cli_non_tty_without_flag_refuses`). Live `--return` (keep_checkout=False) e2e: staged one commit, tree back on `main` clean, printed the exact `git diff`/`git merge`/`git branch -D` commands, and `git branch -D` walked away with the marker gone. Note (non-blocking): `stage_install`'s `non_interactive` param is vestigial (consent lives in `cli_stage`); the engine/consent layering is intentional and correct, so no change required.
- framework/tests/test_install_branch.py 9 passed; full framework suite 879 passed; python bridge+cli (PYTHONPATH=python/src) 126 passed, full python 139 passed; ruff clean on framework/ and python/src+tests.

--- a/python/src/real_team/cli.py
+++ b/python/src/real_team/cli.py
@@ -667,6 +667,17 @@ def install_branch(
         help="After staging, check the original branch back out (tree left clean; install "
         "parked on the branch) instead of staying on the install branch",
     ),
+    expect: str = typer.Option(
+        "any",
+        "--expect",
+        help="repo.expect for the staged bootstrap: fresh|existing|any (default: any — "
+        "install-branch targets an existing repo)",
+    ),
+    config: str = typer.Option(
+        None,
+        "--config",
+        help="Path to a unified YAML install config threaded to the staged bootstrap",
+    ),
 ) -> None:
     """Stage a framework install onto a throwaway git branch — trial it, then merge or delete.
 
@@ -683,6 +694,8 @@ def install_branch(
         target_path,
         branch=branch,
         owner=owner,
+        install_config=config,
+        expect=expect,
         dry_run=dry_run,
         non_interactive=non_interactive,
         return_to_original=return_to_original,

--- a/python/src/real_team/framework_install.py
+++ b/python/src/real_team/framework_install.py
@@ -172,6 +172,7 @@ def stage_install_framework(
     merge_model: str | None = None,
     install_config: Path | str | None = None,
     with_ontology: bool | None = None,
+    expect: str | None = None,
     dry_run: bool = False,
     non_interactive: bool = True,
     return_to_original: bool = False,
@@ -186,9 +187,14 @@ def stage_install_framework(
 
     The staging flags (``branch``/``dry_run``/``non_interactive``/``return_to_original``) are
     consumed by ``install_branch.py``; the install-shape flags (``owner``/``shell``/``merge_model``
-    /… — the same set :func:`install_framework` builds) pass THROUGH it to the bootstrapper that
-    runs on the branch. ``--no-team`` is always forwarded (the CLI owns roster scaffolding), as in
-    :func:`install_framework`.
+    /``expect``/… — the same set :func:`install_framework` builds, plus ``expect``) pass THROUGH
+    it to the bootstrapper that runs on the branch. ``--no-team`` is always forwarded (the CLI owns
+    roster scaffolding), as in :func:`install_framework`.
+
+    ``expect`` forwards ``--expect`` (``fresh``/``existing``/``any``) to the staged bootstrap's
+    fresh-vs-existing gate. Staging by definition targets a repo that already has a HEAD commit
+    (there is nothing to branch from otherwise), so callers should default this to ``"any"`` —
+    the shipped bootstrapper default (``fresh``) refuses every realistic install-branch target.
 
     Returns the CompletedProcess, or None when the framework assets could not be located.
     """
@@ -219,5 +225,7 @@ def stage_install_framework(
         cmd += ["--with-ontology"]
     elif with_ontology is False:
         cmd += ["--no-ontology"]
+    if expect:
+        cmd += ["--expect", expect]
 
     return subprocess.run(cmd, capture_output=True, text=True)

--- a/python/tests/test_cli.py
+++ b/python/tests/test_cli.py
@@ -1558,3 +1558,75 @@ class TestInstallBranch:
         )
         assert branches.stdout.strip() == ""
         assert not (tmp_path / ".claude").exists()
+
+    def test_succeeds_on_realistic_repo(self, tmp_path: Path):
+        """Regression for #283's must-fix: a real repo (HEAD commit + a tracked file) is exactly
+        what install-branch exists to target, but the staged bootstrap used to run with no
+        ``--expect`` and inherit the shipped default ``repo.expect=fresh`` — which the
+        fresh-vs-existing gate REFUSES for a repo with a commit, rolling the staged branch back.
+        The command must default ``--expect`` to ``any`` so this succeeds out of the box."""
+        import subprocess
+
+        subprocess.run(["git", "init", "-q", str(tmp_path)], check=True)
+        (tmp_path / "README.md").write_text("# realistic repo\n", encoding="utf-8")
+        subprocess.run(["git", "-C", str(tmp_path), "add", "README.md"], check=True)
+        subprocess.run(
+            ["git", "-C", str(tmp_path), "-c", "user.name=T", "-c", "user.email=t@t",
+             "commit", "-q", "-m", "seed"], check=True,
+        )
+        result = runner.invoke(
+            app, ["install-branch", "--target", str(tmp_path), "--non-interactive"]
+        )
+        assert result.exit_code == 0, result.output
+        branches = subprocess.run(
+            ["git", "-C", str(tmp_path), "branch", "--list", "real-team/install"],
+            capture_output=True, text=True,
+        )
+        assert "real-team/install" in branches.stdout
+        installed = subprocess.run(
+            ["git", "-C", str(tmp_path), "show", "real-team/install:.claude/framework.config.json"],
+            capture_output=True, text=True,
+        )
+        assert installed.returncode == 0, installed.stderr
+
+    def _capture_cmd(self, monkeypatch) -> dict:
+        import subprocess
+
+        import real_team.framework_install as fi
+
+        captured: dict = {}
+
+        def fake_run(cmd, **kwargs):
+            captured["cmd"] = cmd
+            return subprocess.CompletedProcess(cmd, 0, "", "")
+
+        monkeypatch.setattr(fi.subprocess, "run", fake_run)
+        return captured
+
+    def test_default_threads_expect_any(self, monkeypatch, tmp_path: Path):
+        """Without an explicit --expect, the CLI must default to "any" — install-branch targets
+        an existing repo by definition, and the bootstrapper's shipped default (repo.expect=fresh)
+        would refuse it. subprocess.run is faked here to isolate flag-threading from the (heavy,
+        already-covered-elsewhere) real bootstrap run."""
+        captured = self._capture_cmd(monkeypatch)
+        result = runner.invoke(
+            app, ["install-branch", "--target", str(tmp_path), "--non-interactive"]
+        )
+        assert result.exit_code == 0, result.output
+        cmd = captured["cmd"]
+        assert "--expect" in cmd
+        assert cmd[cmd.index("--expect") + 1] == "any"
+
+    def test_expect_and_config_override_threaded(self, monkeypatch, tmp_path: Path):
+        captured = self._capture_cmd(monkeypatch)
+        cfg = tmp_path / "install.yaml"
+        cfg.write_text("version: 1\n", encoding="utf-8")
+        result = runner.invoke(
+            app,
+            ["install-branch", "--target", str(tmp_path), "--expect", "existing",
+             "--config", str(cfg), "--non-interactive"],
+        )
+        assert result.exit_code == 0, result.output
+        cmd = captured["cmd"]
+        assert "--expect" in cmd and "existing" in cmd
+        assert "--install-config" in cmd and str(cfg) in cmd

--- a/python/tests/test_framework_install.py
+++ b/python/tests/test_framework_install.py
@@ -67,6 +67,26 @@ class TestStageInstallFrameworkBridge:
         assert "--owner" in cmd and "acme" in cmd
         assert "--merge-model" in cmd and "wave-branch" in cmd
 
+    def test_expect_threaded(self, monkeypatch, tmp_path: Path):
+        """``expect`` forwards --expect to the bootstrapper that runs on the staged branch —
+        the fix for #283's must-fix (the staged bootstrap used to inherit the shipped default
+        repo.expect=fresh, which refuses every realistic install-branch target)."""
+        from real_team.framework_install import stage_install_framework
+
+        captured = _capture(monkeypatch)
+        stage_install_framework(tmp_path, expect="existing")
+        cmd = captured["cmd"]
+        assert "--expect" in cmd and "existing" in cmd
+
+    def test_expect_omitted_when_unset(self, monkeypatch, tmp_path: Path):
+        """The bridge itself has no expect default (None): the CLI is the one that defaults to
+        "any"; leaving expect unset here threads no --expect flag at all."""
+        from real_team.framework_install import stage_install_framework
+
+        captured = _capture(monkeypatch)
+        stage_install_framework(tmp_path)
+        assert "--expect" not in captured["cmd"]
+
     def test_no_assets_returns_none(self, monkeypatch, tmp_path: Path):
         import real_team.framework_install as fi
 

